src/replication.c

```diff
@@ -126,37 +126,215 @@ static int handle_mgmt_event_fd(replica_t *replica);
+	mgmt_cmd->data = buf;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	if (write(replica->mgmt_eventfd1, &num, sizeof (num)) !=
+	    sizeof (num)) {
+		REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+		    "for replica(%p)\n", replica);
+		return (ret = -1);
+	}
+	REPLICA_LOG("Start_rebuild opcode sent for Replica ip:%s port:%d "
+	    "state:%d\n", replica->ip, replica->port, replica->state);
+	return ret;
+}
+
+/* Rebuild can be of two type:-
+ * 1. Rebuilding a replica from healthy replica
+ * 2. Rebuilding a replica from all other replica(Mesh rebuild)
+ * API does the following tasks:-
+ * 1. find if there is healthy replica available for rebuilding
+ * 2. find downgraded replica for rebuild
+ * 3. Call prepare_for_rebuild to on healthy replica or on all
+ * other replicas to get IP address and port for replicas.
  */
 void
 update_volstate(spec_t *spec)
 {
+	int ret = 0;
 	uint64_t max;
-	replica_t *replica;
+	replica_t *replica = NULL;
+	replica_t *target_replica = NULL;
+	replica_t *healthy_replica = NULL;
+
+	if (spec->rebuild_in_progress == true) {
+		REPLICA_NOTICELOG("Rebuild is already in progress "
+		    "on replica:%s\n", spec->volname);
+		return;
+	}
+
+	if (((spec->healthy_rcount + spec->degraded_rcount >=
+	    spec->consistency_factor) && (spec->healthy_rcount >= 1)) ||
+	    (spec->healthy_rcount  + spec->degraded_rcount >=
+	    MAX(spec->replication_factor - spec->consistency_factor + 1,
+	    spec->consistency_factor))) { 
 
-	if(((spec->healthy_rcount + spec->degraded_rcount >= spec->consistency_factor) &&
-		(spec->healthy_rcount >= 1))||
-		(spec->healthy_rcount  + spec->degraded_rcount 
-			>= MAX(spec->replication_factor - spec->consistency_factor + 1, spec->consistency_factor))) {
-		if (spec->ready == false)
-		{
-			max = 0;
-			TAILQ_FOREACH(replica, &spec->rq, r_next)
-				max = (max < replica->initial_checkpointed_io_seq) ?
-				    replica->initial_checkpointed_io_seq : max;
+		max = 0;
+		TAILQ_FOREACH(replica, &spec->rq, r_next) {
+			/* Find healthy replica */
+			if ((healthy_replica == NULL) &&
+			    (replica->state == ZVOL_STATUS_HEALTHY)) {
+				healthy_replica = replica;
+			}
+
+			/*
+			 * If we know master replica and healthy
+			 * replica then trigger rebuild
+			 */
+			if ((spec->master_replica != NULL) &&
+			    (healthy_replica != NULL))
+				break;
+			/*
+			 * Should not account io_seq from healthy replica
+			 * when trying to find out degraded replica
+			 */
+			if (replica->state == ZVOL_STATUS_HEALTHY)
+				continue;
+			if (max < replica->initial_checkpointed_io_seq) {
+				max = replica->initial_checkpointed_io_seq;
+				target_replica = replica;
+			}
+		}
 
+		/*
+		 * Increase io_seq only when we do not have healthy replica
+		 */	
+		if (healthy_replica == NULL) {
 			max = (max == 0) ? 10 : max + (1<<20);
 			spec->io_seq = max;
```

Comment:
may not be related to this PR, spec's io_seq need to modified only when istgt is restarted and spec is turning 'ready' for first time. Any successive state change of spec doesn't need to update io_seq.

Reply:
And, if any healthy replica is there, just taking its check_pointed_io_seq and incrementing it should be good enough.

Reply:
regarding to setting io_seq (as algorithm), scenarios I'm thinking are of:
case 1: istgt is restarted, and, volume is becoming 'not ready' to 'ready state, after required number of replicas are connected in degraded state.
case 2: istgt is running. It went from 'ready' state to 'not ready' state due to replicas disconnection. It again became 'ready', but, there are no healthy replicas.
case 3: istgt is running. It went from 'ready' state to 'not ready' state due to replicas disconnection. It again became 'ready', with atleast a healthy replica in it.
For second and third cases, we don't need to change io_seq, as istgt already knows the io_seq, and it don't wrote any other IOs.
For first case, looking out for max, and adding (1<<20) should be fine.
As you mentioned, healthy replica case is not needed to handle. I mentioned that comment wrt case 3, and, check_pointed_io_seq being old than the latest one.

Reply:
regarding to this PR, we are not looking for transition from 'not ready' to 'ready'. This can lead to regular updates of io_seq. Is check for 'spec->ready' being false removed intentionally?
Combining spec's 'ready' with rebuilding seems to be making this logic complex. If things can go hand in hand, it should be fine leaving them together.